src/main/java/it/gov/pagopa/pu/pagopapayments/connector/auth/service/AuthAccessTokenRetriever.java

```diff
@@ -3,51 +3,54 @@
 import org.springframework.beans.factory.annotation.Value;
 import org.springframework.context.annotation.Lazy;
 import org.springframework.stereotype.Service;
 
 import java.time.LocalDateTime;
-import java.util.concurrent.atomic.AtomicReference;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 
 @Slf4j
 @Lazy
 @Service
 public class AuthAccessTokenRetriever {
 
-    private static final String GRANT_TYPE = "client_credentials";
-    private static final String SCOPE = "openid";
-    private static final String CLIENT_ID_PREFIX = "piattaforma-unitaria_";
-
-    private final AuthnClient authnClient;
-    private final String clientSecret;
-
-    private final AtomicReference<Pair<LocalDateTime, AccessToken>> accessTokenRef = new AtomicReference<>();
-
-    public AuthAccessTokenRetriever(
-            @Value("${rest.auth.post-token.client_secret}")
-            String clientSecret,
-
-            AuthnClient authnClient) {
-        this.authnClient = authnClient;
-        this.clientSecret = clientSecret;
-    }
-
-    public AccessToken getAccessToken() {
-        return accessTokenRef.updateAndGet(this::checkAndReturn).getValue();
-    }
-
-    private Pair<LocalDateTime, AccessToken> checkAndReturn(Pair<LocalDateTime, AccessToken> tokenPair) {
-        if (tokenPair == null || LocalDateTime.now().isAfter(tokenPair.getLeft())) {
-            log.info("M2M AccessToken expired, refreshing");
-            LocalDateTime tokenRequestDateTime = LocalDateTime.now();
-            AccessToken accessToken = authnClient.postToken(CLIENT_ID_PREFIX, GRANT_TYPE, SCOPE, null, null, null, clientSecret);
-            LocalDateTime expiration = tokenRequestDateTime.plusSeconds(accessToken.getExpiresIn() - 5L); // setting some seconds to avoid too strict expiration
-            return Pair.of(expiration, accessToken
-            );
-        } else {
-            return tokenPair;
-        }
-    }
+  private static final String GRANT_TYPE = "client_credentials";
+  private static final String SCOPE = "openid";
+  private static final String CLIENT_ID_PREFIX = "piattaforma-unitaria_";
+  private static final String NULL_ORG_ID = "!NULL_ORG_IPA_CODE!";
+
+  private final AuthnClient authnClient;
+  private final String clientSecret;
+
+  private final Map<String, Pair<LocalDateTime, AccessToken>> accessTokenRefMap = new ConcurrentHashMap<>();
+
+  public AuthAccessTokenRetriever(
+    @Value("${rest.auth.post-token.client_secret}")
+    String clientSecret,
+
+    AuthnClient authnClient) {
+    this.authnClient = authnClient;
+    this.clientSecret = clientSecret;
+  }
+
+  public AccessToken getAccessToken(String orgIpaCode) {
+    String orgIpaCodeKey = orgIpaCode == null ? NULL_ORG_ID : orgIpaCode;
+    return accessTokenRefMap.compute(orgIpaCodeKey, (k, v) -> {
+      if (v == null || LocalDateTime.now().isAfter(v.getLeft())) {
+        String clientId = CLIENT_ID_PREFIX + StringUtils.stripToEmpty(orgIpaCode);
```

Comment:
if we store as key the cliendIt instead?
```suggestion
    String clientId = CLIENT_ID_PREFIX + StringUtils.stripToEmpty(orgIpaCode);
    return accessTokenRefMap.compute(clientId, (clientId, v) -> {
      if (v == null || LocalDateTime.now().isAfter(v.getLeft())) {
```

Reply:
done